### PR TITLE
[REVIEW] Next UI Hackathon

### LINKED ISSUE #3691: UI Overhaul: Sprint 50 Hackathon
```[tasklist]
### Tasks
- [x] Make detail view look like Figma design @bryanculver @HanlinMiao
- [x] Pull tab data from API @HanlinMiao
- [x] Make list view look like Figma design @timizuoebideri1
- [x] Move all view slug URLs to PK URLs @gsnider2195
- [x] Polish up Detail endpoint @HanlinMiao
- [x] Pagination @HanlinMiao
- [x] Improve accordion UX @glennmatthews
- [x] Create homepage view @glennmatthews
- [x] Default columns for list view (defined on model) @jathanism
- [x] Improve home view loading performance @gsnider2195
- [x] Find something plugin-centric to be included into the UI @glennmatthews
- [ ] Get edit view loading from an API-provided schema @jathanism
- [x] Improve home view grid and styling
- [x] Move `depth=0` API to return walkable response (hyperlink) @glennmatthews #3679
- [x] **nautobot-ui**: Fix horizontal scrolling going off the view not working for non-Firefox browsers @norbert-mieczkowski-codilime
- [ ] **nautobot-ui**: Fixed columns (checkbox, name?, action menu) @norbert-mieczkowski-codilime
- [ ] Move all queries over to RTK pattern (POST + invalidate GET) @bryanculver
- [x] Cache/local storage quirks
- [ ] Remove M2M fields from `depth=0` response @glennmatthews #3702
```

### PR DESCRIPTION
**Pull request by bryanculver**
Closes #3691

### CODE REVIEW

File: nautobot/core/apps/__init__.py [hunk 1]
```diff
@@ -94,43 +106,65 @@ def register_menu_items(tab_list):
                         f"{nav_tab.name} -> {group.name} -> {item.link}",
```
Comment (glennmatthews) [inline nautobot/core/apps/__init__.py:106]: It's a bit awkward data-structure wise that we index the leaf menu items in the registry by their `link` rather than their `name` when the rest of the structure is indexed by `name`. Might consider changing this as it may allow for some code consolidation.

File: nautobot/core/apps/__init__.py [hunk 2]
```diff
@@ -94,43 +106,65 @@ def register_menu_items(tab_list):
...
+            sorted(registry_groups.items(), key=lambda kv_pair: kv_pair[1]["weight"])
+        )
+        # Add collected permissions to tab dict
+        registry["nav_menu"][nav_tab.name]["permissions"] |= tab_perms
 
         # Order all tabs in dict
-        registry["nav_menu"]["tabs"] = OrderedDict(
-            sorted(registry["nav_menu"]["tabs"].items(), key=lambda kv_pair: kv_pair[1]["weight"])
+        registry["nav_menu"] = OrderedDict(
+            sorted(registry["nav_menu"].items(), key=lambda kv_pair: kv_pair[1]["weight"])
```
Comment (glennmatthews) [inline nautobot/core/apps/__init__.py:167]: FYI this change here to reduce the nesting by one level is why I changed `extras.registry` to allow the top-level keys of the registry to be mutable.

It might be better in revisiting this to revert that change and simply initialize `registry["nav_menu"]` with the fixed ordered list of keys defined in `MENU_TABS` now.

File: nautobot/core/apps/__init__.py [hunk 3]
```diff
@@ -483,44 +531,44 @@ def __init__(self, name, items=None, weight=1000):
...
             "args": [],
             "kwargs": {},
         }
 
     @property
-    def fixed_fields(self):
+    def fixed_fields(self) -> tuple:
+        """Tuple of (name, attribute) entries describing fields that may not be altered after declaration."""
         return (
-            ("name", self.name),
```
Comment (glennmatthews) [inline nautobot/core/apps/__init__.py:514]: Not sure why `name` was removed here - might have been in error?

File: nautobot/core/models/validators.py [hunk 4]
```diff
@@ -50,3 +50,7 @@ def validate_regex(value):
         re.compile(value)
     except re.error:
         raise ValidationError(f"{value} is not a valid regular expression.")
+
+
+# FIXME(jathan): Mke this a real validator subclass?
+validate_regex.message = "Jathan was here."
```
Comment (glennmatthews) [inline nautobot/core/models/validators.py:56]: TODO :-)

File: nautobot/core/templates/plugin_template/navigation.py-tpl
Comment (glennmatthews) [inline nautobot/core/templates/plugin_template/navigation.py-tpl:1]: We should remove the entire plugin_template in favor of cookiecutter, but that's out of scope for this PR.

File: nautobot/extras/api/views.py [hunk 5]
```diff
@@ -346,6 +368,74 @@ def app_full_width_fragment(self, request, pk):
...
+    @action(detail=True, url_path="detail_view_config")
+    def detail_view_config(self, request, pk):
+        """
+        Return a JSON of the ObjectDetailView configuration
+        """
+        obj = get_object_or_404(self.queryset, pk=pk)
+        obj_serializer_class = get_serializer_for_model(obj)
+        obj_serializer = obj_serializer_class(data=None)
+        response = self.get_detail_view_config(obj_serializer)
+        response = JsonResponse(response)
```
Comment (glennmatthews) [inline nautobot/extras/api/views.py:380]: Should we be using `rest_framework.Response` here instead for consistency?

File: nautobot/extras/api/views.py [hunk 6]
```diff
@@ -191,6 +191,28 @@ def table_fields(self, request):
         ]
         return Response({"data": data})
 
+    @action(detail=False, url_path="note-table-fields", methods=["get"])
+    def note_table_fields(self, request):
```
Comment (glennmatthews) [inline nautobot/extras/api/views.py:195]: docstring please

File: nautobot/extras/api/views.py [hunk 7]
```diff
@@ -191,6 +191,28 @@ def table_fields(self, request):
...
+        table = get_table_for_model(Note)
+        table_instance = table(user=request.user, data=[])
+        data = [
+            {"name": item[0], "label": item[1]}
+            for item in table_instance.configurable_columns
+            if item[0] in table_instance.visible_columns
+        ]
+        return Response({"data": data})
+
+    @action(detail=False, url_path="changelog-table-fields", methods=["get"])
```
Comment (glennmatthews) [inline nautobot/extras/api/views.py:205]: docstring please

File: nautobot/extras/api/views.py [hunk 8]
```diff
@@ -346,6 +368,74 @@ def app_full_width_fragment(self, request, pk):
...
+        Return a JSON of the ObjectDetailView configuration
+        """
+        obj = get_object_or_404(self.queryset, pk=pk)
+        obj_serializer_class = get_serializer_for_model(obj)
+        obj_serializer = obj_serializer_class(data=None)
+        response = self.get_detail_view_config(obj_serializer)
+        response = JsonResponse(response)
+        return response
+
+    def get_detail_view_config(self, obj_serializer):
```
Comment (glennmatthews) [inline nautobot/extras/api/views.py:383]: docstring please

File: nautobot/extras/plugins/__init__.py [hunk 9]
```diff
@@ -515,124 +511,17 @@ def register_filter_extensions(filter_extensions, plugin_name):
 #
 
 
-# 2.2 TODO: remove in favor of NavMenuItem
-@class_deprecated_in_favor_of(NavMenuItem)
-class PluginMenuItem:
```
Comment (glennmatthews) [inline nautobot/extras/plugins/__init__.py:520]: Need to document as removed

File: nautobot/extras/registry.py [hunk 10]
```diff
@@ -14,8 +14,8 @@ def __getitem__(self, key):
             raise KeyError(f"Invalid store: {key}")
 
     def __setitem__(self, key, value):
-        if key in self:
-            raise KeyError(f"Store already set: {key}")
+        # if key in self:
+        #     raise KeyError(f"Store already set: {key}")
         super().__setitem__(key, value)
```
Comment (glennmatthews) [inline nautobot/extras/registry.py:19]: This was to make the root keys of the registry overwritable - see earlier comment about how it might be feasible to revert this and handle the nav menu root entries differently.

File: nautobot/extras/tests/test_registry.py [hunk 11]
```diff
@@ -23,12 +23,6 @@ def test_manipulate_store(self):
 
         self.assertListEqual(self.reg["foo"], [1, 2, 3])
 
-    def test_overwrite_store(self):
-        self.reg["foo"] = 123
-
-        with self.assertRaises(KeyError):
-            self.reg["foo"] = 456
-
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_registry.py:31]: Could revert this if we revert the change to `registry.__setitem__`

File: nautobot/ui/craco.config.js [hunk 12]
```diff
@@ -1,14 +1,16 @@
-const { CracoAliasPlugin } = require("react-app-alias-ex");
+const {
+    NautobotCracoAliasPlugin,
+} = require("./lib/nautobot-craco-alias-plugin");
 
 // Suppress console logging unless debugging is explicitly enabled.
 if (!process.env.NAUTOBOT_DEBUG) {
-    console.log = function () {};
+    // console.log = function () {};
```
Comment (glennmatthews) [inline nautobot/ui/craco.config.js:7]: Revert?

File: nautobot/ui/lib/nautobot-craco-alias-plugin.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
+/**
+ * This is a wrapper around react-app-alias and react-app-alias-ex that addresses
+ * a bug in react-app-alias-ex where it does not properly resolve the path to
+ * node_modules when running Jest tests.
+ *
+ * See: https://github.com/oklas/react-app-alias/issues/95
```
Comment (glennmatthews) [inline nautobot/ui/lib/nautobot-craco-alias-plugin.js:6]: @bryanculver your latest comments on this issue make it sound like #3618 may have removed the need for this wrapper/workaround - still needed?

File: nautobot/ui/src/components/ObjectListTable.js [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+        // TODO: ui-schema should be providing the name of the field used to describe the object (name)
+        return window.location.pathname + obj.id;
+    }
+    if (typeof obj !== "object" || !obj || !obj.url) {
+        return null;
+    }
+    // Remove domain + /api prefix
+    const url = obj.url.replace(window.location.origin + "/api", "");
+
+    // Statuses and Roles should not be linkable
```
Comment (glennmatthews) [inline nautobot/ui/src/components/ObjectListTable.js:31]: Why?

File: nautobot/ui/src/components/ObjectListTable.js [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+    if (
+        ["/extras/statuses", "/extras/roles"].some((prefix) =>
+            url.startsWith(prefix)
```
Comment (glennmatthews) [inline nautobot/ui/src/components/ObjectListTable.js:34]: url always starts with `/api` if I'm reading the above right, so this seems like dead code anyways?

File: nautobot/ui/src/components/SidebarNav.js [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+                                                            menu[1].items
+                                                        ).map(
+                                                            (
+                                                                submenu,
+                                                                submenu_idx,
+                                                                submenu_arr
+                                                            ) => (
+                                                                <SidebarButton
+                                                                    as={
+                                                                        ReactRouterLink
```
Comment (glennmatthews) [inline nautobot/ui/src/components/SidebarNav.js:106]: Switching to a 2-space indent and/or splitting some of this out into subcomponents would greatly help readability IMO.

File: nautobot/ui/src/utils/store.js [hunk 17]
```diff
@@ -19,13 +20,106 @@ const persistConfig = {
...
+    reducers: {
+        updateAppContext(state, action) {
+            state = action.payload;
+            return state;
+        },
+    },
+});
+
+const initialAppState = {
+    currentContext: "Inventory", // TODO: What is the context for the homepage?
```
Comment (glennmatthews) [inline nautobot/ui/src/utils/store.js:37]: Should be empty and the sidebar-nav shouldn't be shown on the homepage, IMO.

File: nautobot/core/api/metadata.py [hunk 18]
```diff
@@ -1,14 +1,19 @@
...
+from rest_framework import fields as drf_fields
+from rest_framework import serializers as drf_serializers
 from rest_framework.metadata import SimpleMetadata
 from rest_framework.request import clone_request
 
 from nautobot.core.api import ContentTypeField
 
 
-class NautobotMetadata(SimpleMetadata):
+class NautobotMetadataV1(SimpleMetadata):
```
[the remark this answers is not in the document]
Comment (jathanism) [reply nautobot/core/api/metadata.py:16]: Yes.

File: nautobot/core/api/metadata.py [hunk 19]
```diff
@@ -57,3 +62,184 @@ def get_field_info(self, field):
...
+                result[k] = result_default[k]
+        return result
+
+
+class NautobotSchemaProcessor(NautobotProcessingMixin, SchemaProcessor):
+    """
+    SchemaProcessor to account for custom field types and behaviors for Nautobot.
+    """
+
+    def _get_field_properties(self, field: SerializerType, name: str) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (jathanism) [reply nautobot/core/api/metadata.py]: The challenge is adding support for `ManyRelatedField` for nested child relationships. Let's revisit next week, because I've had to make a handful of other changes to adhere properly to JSON schema and we should discuss strategy here.

File: nautobot/core/api/metadata.py [hunk 20]
```diff
@@ -57,3 +62,184 @@ def get_field_info(self, field):
...
+        result = {}
+        type_map_obj = self._get_type_map_value(field)
+        result["type"] = type_map_obj["type"]
+        result["title"] = self._get_title(field, name)
+
+        if isinstance(field, drf_serializers.ListField):
+            if field.allow_empty:
+                result["required"] = not getattr(field, "allow_empty", True)
+            result["items"] = self._get_field_properties(field.child, "")
+            result["uniqueItems"] = True
```
[the remark this answers is not in the document]
Comment (jathanism) [reply nautobot/core/api/metadata.py:103]: The challenge is adding support for `ManyRelatedField` for nested child relationships. Let's revisit next week, because I've had to make a handful of other changes to adhere properly to JSON schema and we should discuss strategy here.

File: nautobot/core/api/metadata.py [hunk 21]
```diff
@@ -57,3 +62,184 @@ def get_field_info(self, field):
...
+            fields = self.serializer.child.get_fields()
+        else:
+            fields = self.serializer.get_fields()
+
+        field_names = list(fields)
+        bottom_fields = ["computed_fields", "custom_fields", "relationships"]
+        for field_name in bottom_fields:
+            if field_name in field_names:
+                field_names.remove(field_name)
+                field_names.append(field_name)
```
[the remark this answers is not in the document]
Comment (jathanism) [reply nautobot/core/api/metadata.py:167]: Yeah it's ghetto. I was just iterating to get something working. I want a central solution for this.

Agreed.

File: nautobot/core/api/metadata.py [hunk 22]
```diff
@@ -57,3 +62,184 @@ def get_field_info(self, field):
...
+
+    - schema: The object JSON schema
+    - uiSchema: The object UI schema which describes the form layout in the UI
+    """
+
+    def _determine_actions(self, request, view):
+        """
+        Replace the stock determine_actions() method to assess object permissions only
+        when viewing a specific object. This is necessary to support OPTIONS requests
+        with bulk update in place (see #5470).
```
[the remark this answers is not in the document]
Comment (jathanism) [reply nautobot/core/api/metadata.py]: Makes me wonder if this is even still a thing for v2 with the changes we're making.

File: nautobot/ui/src/components/paginator_form.js [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+            <Select
+                value={
+                    searchParams.get("limit") ? searchParams.get("limit") : "50"
+                }
+                onChange={onPageSizeChange}
+            >
+                <option value="50">50</option>
+                <option value="100">100</option>
+                <option value="200">200</option>
+                <option value="500">500</option>
```
Comment (glennmatthews) [inline nautobot/ui/src/components/paginator_form.js:28]: TODO: we need a REST API endpoint to query `get_settings_or_config("PER_PAGE_DEFAULTS")` rather than hard-coding this.

File: nautobot/ui/src/utils/store.js [hunk 24]
```diff
@@ -19,13 +20,105 @@ const persistConfig = {
...
+                            let tokens = urlPatternOrSubGroup.split("/");
+                            if (tokens.length === 4) {
+                                let appLabel = tokens[1];
+                                let modelNamePlural = tokens[2];
+                                if (appLabel in urlPatternToContext === false) {
+                                    urlPatternToContext[appLabel] = {};
+                                }
+                                urlPatternToContext[appLabel][modelNamePlural] =
+                                    context;
+                            }
```
Comment (glennmatthews) [inline nautobot/ui/src/utils/store.js:70]: This probably needs to be updated to handle plugin routes correctly.

File: nautobot/ui/src/views/Home.js [hunk 25]
```diff
@@ -1,11 +1,144 @@
...
+    Thead,
+    Tr,
+} from "@nautobot/nautobot-ui";
+import HomeChangelogPanel from "@components/HomeChangelogPanel";
+import HomePanel from "@components/HomePanel";
+import JobHistoryTable from "@components/JobHistoryTable";
+import { LoadingWidget } from "@components/LoadingWidget";
+import { useGetRESTAPIQuery } from "@utils/api";
+import GenericView from "@views/generic/GenericView";
+import { useGetObjectCountsQuery } from "@utils/api";
```
Comment (glennmatthews) [inline nautobot/ui/src/views/Home.js:30]:
```suggestion
import { useGetObjectCountsQuery, useGetRESTAPIQuery } from "@utils/api";
import GenericView from "@views/generic/GenericView";
```

File: nautobot/ui/src/views/Home.js [hunk 26]
```diff
@@ -1,11 +1,144 @@
...
+                icon=<SecurityIcon />
+                title="Security"
+                data={[
+                    { name: "Menu Item 1", count: 0 },
+                    { name: "Menu Item 2", count: 0 },
+                    { name: "Menu Item 3", count: 0 },
+                    { name: "Menu Item 4", count: 0 },
+                    { name: "Menu Item 5", count: 0 },
+                    { name: "Menu Item 6", count: 0 },
+                ]}
```
Comment (glennmatthews) [inline nautobot/ui/src/views/Home.js:83]: TODO: this should use `objectCountData["Security"]`

File: nautobot/core/api/views.py [hunk 27]
```diff
@@ -821,34 +701,149 @@ def execute_graphql_request(self, request, data, query, variables, operation_nam
...
+                {"model": "dcim.virtualchassis"},
+                {"model": "dcim.deviceredundancygroup"},
+                {"model": "dcim.cable"},
+            ],
+            "Networks": [
+                {"model": "ipam.vrf"},
+                {"model": "ipam.prefix"},
+                {"model": "ipam.ipaddress"},
+                {"model": "ipam.vlan"},
+            ],
```
Comment (glennmatthews) [inline nautobot/core/api/views.py:813]: Let's add `Security: [{"model": "extras.secret"}]` here just so we have some data for now.

File: nautobot/core/api/views.py [hunk 28]
```diff
@@ -821,34 +701,149 @@ def execute_graphql_request(self, request, data, query, variables, operation_nam
...
+        return Response(filtered_menu)
+
+
+class GetObjectCountsView(NautobotAPIVersionMixin, APIView):
+    """
+    Enumerate the models listed on the Nautobot home page and return data structure
+    containing verbose_name_plural, url and count.
+    """
+
+    permission_classes = [IsAuthenticated]
```
[the remark this answers is not in the document]
Comment (gsnider2195) [reply nautobot/core/api/views.py:795]: I was thinking we could return a null count from the API and have the UI show the lock icon

File: nautobot/core/api/views.py [hunk 29]
```diff
@@ -821,34 +701,149 @@ def execute_graphql_request(self, request, data, query, variables, operation_nam
...
+            app_label, model_name = entry["model"].split(".")
+            model = apps.get_model(app_label, model_name)
+            permission = get_permission_for_model(model, "view")
+            if HIDE_RESTRICTED_UI and not request.user.has_perm(permission):
+                continue
+            data = {"name": model._meta.verbose_name_plural}
+            try:
+                data["url"] = django_reverse(get_route_for_model(model, "list"))
+            except NoReverseMatch:
+                pass
```
[the remark this answers is not in the document]
Comment (gsnider2195) [reply nautobot/core/api/views.py:837]: It doesn't with our curated list of models but when I was using all models some of them didn't have a list view which is why this was added. I'd say lets just remove the try/except until we decide to add more models.

File: nautobot/ui/src/components/paginator_form.js [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+    function onPageSizeChange(event) {
+        let offset = searchParams.get("offset");
+        setSearchParams({
+            limit: event.target.value,
+            offset: offset ? offset : 0,
+        });
+    }
+
+    return (
+        <Box width="200px">
```
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/ui/src/components/paginator_form.js:18]: without `width`:
Screen Shot 2023-05-02 at 3 07 32 PM

with `width`:
Screen Shot 2023-05-02 at 3 07 45 PM

File: nautobot/ui/src/views/generic/ObjectRetrieve.js [hunk 31]
```diff
@@ -1,53 +1,125 @@
...
+//     });
+
+function render_header(value) {
+    value = toTitleCase(value, "_");
+    value = toTitleCase(value, "-");
+    return value;
+}
 
-function render_value(value) {
+function Render_value(value) {
```
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/ui/src/views/generic/ObjectRetrieve.js:71]: Screen Shot 2023-05-02 at 3 12 53 PM

File: nautobot/ui/src/components/paginator.js [hunk 32]
```diff
@@ -0,0 +1,75 @@
+import { Frame } from "@nautobot/nautobot-ui";
+import React from "react";
+
+import PaginatorForm from "@components/paginator_form";
+import Pagination from "@components/pagination";
+
+import { NautobotGrid, NautobotGridItem } from "@nautobot/nautobot-ui";
+
+class Paginator extends React.Component {
```
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/ui/src/components/paginator.js:9]: I think the only difference I see so far is that you **cannot** use a `ReactHook` within a `React.Component` class, while you can in a `function`

File: nautobot/ui/src/views/generic/ObjectRetrieve.js (hunk 31 above)
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/ui/src/views/generic/ObjectRetrieve.js]: Ah okay. Can we call it something like `DetailFieldValue` or similar then?

File: nautobot/core/api/views.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (gsnider2195) [reply nautobot/core/api/views.py:792]: I added the redirect to `/login/` if the user is not authenticated.

File: nautobot/core/api/views.py [hunk 33]
```diff
@@ -822,33 +696,58 @@ def execute_graphql_request(self, request, data, query, variables, operation_nam
 
 
 class GetMenuAPIView(NautobotAPIVersionMixin, APIView):
-    """API View that returns the registered nav-menu content."""
+    """API View that returns the nav-menu content applicable to the requesting user."""
 
-    permission_classes = [AllowAny]
+    permission_classes = [AllowAny]  # TODO: should be IsAuthenticated but this breaks the landing page if logged out
```
[the remark this answers is not in the document]
Comment (gsnider2195) [reply nautobot/core/api/views.py]: With the automatic redirect to `/login/` for unauthenticated users this workaround is no longer required so I changed it to `IsAuthenticated`

File: nautobot/core/apps/__init__.py [hunk 34]
```diff
@@ -56,82 +58,86 @@ def create_or_check_entry(grouping, record, key, path):
...
+            raise TypeError(f"Top level objects need to be an instance of NavMenuTab: {nav_tab}")
+        if nav_tab.name not in MENU_TABS:
+            raise ValueError(f"Unexpected NavMenuTab name: {nav_tab.name}")
+
+        tab_perms = registry["nav_menu"][nav_tab.name]["permissions"]
+        registry_groups = registry["nav_menu"][nav_tab.name]["groups"]
+        # TODO: allow for recursive (more than two-level) nesting of groups?
+        for group in nav_tab.groups:
+            if not isinstance(group, NavMenuGroup):
+                raise TypeError(f"Expected a NavMenuGroup, but got {group}")
```
Comment (gsnider2195) [inline nautobot/core/apps/__init__.py:85]:
```suggestion
                raise TypeError(f"Expected a NavMenuGroup, but got {type(group)}")
```

File: nautobot/core/apps/__init__.py [hunk 35]
```diff
@@ -56,82 +58,86 @@ def create_or_check_entry(grouping, record, key, path):
...
+
+            create_or_check_entry(registry_groups, group, group.name, f"{nav_tab.name} -> {group.name}")
+
+            group_perms = registry_groups[group.name]["permissions"]
+            for item in group.items:
+                if not isinstance(item, (NavMenuGroup, NavMenuItem)):
+                    raise TypeError(f"Expected NavMenuItem or NavMenuGroup, but found {item}")
```
Comment (gsnider2195) [inline nautobot/core/apps/__init__.py:92]:
```suggestion
                    raise TypeError(f"Expected NavMenuItem or NavMenuGroup, but got {type(item)}")
```

File: nautobot/dcim/tests/test_api.py [hunk 36]
```diff
@@ -2140,9 +2140,10 @@ def test_remove_chassis_from_master_device(self):
         virtual_chassis_1 = response.json()["results"][0]
 
         # Make sure the master is set
-        self.assertNotEqual(virtual_chassis_1["master"], None)
+        self.assertIsNotNone(virtual_chassis_1["master"])
 
-        master_device = Device.objects.get(pk=virtual_chassis_1["master"])
+        # The `master` key will be a URL now, but it contains the PK
+        master_device = Device.objects.get(pk=virtual_chassis_1["master"].split("/")[-2])
```
Comment (gsnider2195) [inline nautobot/dcim/tests/test_api.py:2146]: TODO: helper function to extract a pk from an api url or even return the instance of the object?

File: nautobot/core/api/mixins.py [hunk 37]
```diff
@@ -62,6 +63,12 @@ def get_queryset_filter_params(self, data, queryset):
         if isinstance(data, dict):
             params = dict_to_filter_params(data)
             return self.remove_non_filter_fields(params)
+
+        # Account for the fact that HyperlinkedIdentityFields might pass in URLs.
```
Comment (bryanculver) [inline nautobot/core/api/mixins.py:67]: TODO: Let's account for this in the docstrings

File: nautobot/core/settings.py [hunk 38]
```diff
@@ -805,3 +805,35 @@ def UI_RACK_VIEW_TRUNCATE_FUNCTION(device_display_name):
...
+    # Foreign Key fields
+    #
+    # enum=choices is the one that works in the UI as a related field but it
+    # includes ALL related objects in the schema.
+    # "NautobotHyperlinkedRelatedField": {"type": "string", "enum": "choices"},
+    # readOnly=True disables the fields in the UI; not what we want.
+    # "NautobotHyperlinkedRelatedField": {"type": "string", "readOnly": True},
+    # type=string results in a free text field; also not what we want. For now,
+    # however, this will keep things moving so the unit tests pass.
+    "NautobotHyperlinkedRelatedField": {"type": "string", "format": "uuid"},
```
Comment (bryanculver) [inline nautobot/core/settings.py:833]: TODO: Should this be `"format": "url"`?

File: nautobot/ui/src/components/paginator_form.js [hunk 39]
```diff
@@ -0,0 +1,33 @@
...
+            <Select
+                value={
+                    searchParams.get("limit") ? searchParams.get("limit") : "50"
```
[the remark this answers is not in the document]
Comment (bryanculver) [reply nautobot/ui/src/components/paginator_form.js:20]: Should be part of the profile information on first load.

File: nautobot/core/settings.py (hunk 38 above)
Comment (gsnider2195) [reply nautobot/core/settings.py:833]: Looks like it would be [`uri`](https://rjsf-team.github.io/react-jsonschema-form/docs/usage/widgets/#string-formats) but we should probably hold off on making any changes that assume we're going to keep using rjsf

File: nautobot/ui/lib/nautobot-craco-alias-plugin.js (hunk 13 above)
Comment (bryanculver) [reply nautobot/ui/lib/nautobot-craco-alias-plugin.js:6]: Yes, @gsnider2195 still ran into an issue with this when getting the code output from Jest. So it worked for the tests but couldn't decompile. FWIW, the reference issue upstream has been closed with the patch merged in.

File: nautobot/core/settings.py (hunk 38 above)
Comment (glennmatthews) [reply nautobot/core/settings.py:833]: It can be a UUID on write.

**Review by glennmatthews — APPROVED** [review]
Let's merge and iterate!
